### ISSUE TITLE
modifies Relationship or: How I Learned to Stop Worrying and Love Overwrites

### ISSUE THREAD

**Issue by dbent** [KSP-CKAN/CKAN#1311]
So PRs like KSP-CKAN/NetKAN#1916 and KSP-CKAN/NetKAN#1787 make me very :sob:. Why? Because whilst our metadata is robust enough to handle such cases, the metadata becomes *very* fragile; subsequent updates can easily break installations in subtle ways that are hard to identify. Not good for us, not good for users, not good for mod authors.

In an ideal world no mod would overwrite any other mod's files, ever. And in a lot of cases, it's easy to modify mods to make it so it's not even required: valerian/ksp-planetshine#1, BryceSchroeder/Kopernicus#43.

However in the cases where it's not so easy we get things like the above and they pop up with alarming frequency.

My proposal: add a `modifies` relationship.
```json
{
    "spec_version": "1.x",
    "identifier": "NotSoAwesomeMod",
    "modifies": [
        { "name": "AwesomeMod" }
    ],
    "install": [
        {
            "file": "AwesomeMod/AwesomeTexture.dds",
            "install_to": "GameData/AwesomeMod"
        }
    ]
}
```

`modifies` would be similar to `depends`, *except* a mod with a `modifies` relationship would be allowed to overwrite any files owned by the modified mod. The modified mod would **not** be treated as installed, but the `modifies` mod would automatically `conflicts` and `provides` the modified mod.

Installation/Upgrade would work as follows:
- Upon installing/upgrading a mod, check if it has `modifies` relationships.
- If it does, for each mod in the `modifies` relationship, remove them if they are already installed.
- For each mod mod in the `modifies` relationship, install the latest appropriate version depending on version constraints.
- Install the actual mod, allowing files to be overwritten so long as they are owned by a mod in the `modifies` relationship.

Removal would work as follows:
- Upon removing a mod, check if has `modifies` relationships.
- If it does, for each mod in the `modifies` relationship, remove them.
- Remove the actual mod, ignoring files it owns that are now missing (since they were presumably removed in the previous step)

The two biggest issues are:
- Increased relationship complexity, meaning more bugs, harder to reason about, etc.
- Installation ordering is now a thing, meaning circular ~~dependencies~~ *modifications* should be detected and trigger an exception.

If this is implemented we should treat it as a *last resort* option and should encourage mod authors to avoid overwrites as much as possible.

Tagging @plague006 as this might interest him.

**Comment by dbent** [comment]
Since it wasn't explicit in the original post, because of the implicit `conflicts` on the original mod, two mods which `modifies` the same would not be allowed to be installed at the same time.

This is reasonable.

**Comment by plague006** [comment]
I like it overall because the "fragile" relationships I've built lately are *very* easily broken: KSP-CKAN/NetKAN#1916 actually broke 2 times in 3 days due to updates.

I'll throw up another potential issue:

If mod X `depends` mod A which is modified by mod B, then X can't depend on mod A, and we can't *actually* provide mod A since we can't assure which files are present. So we would also need a way to resolve those conflicts.

**Comment by dbent** [comment]
> If mod X depends mod A which is modified by mod B, then X can't depend on mod A, and we can't actually provide mod A since we can't assure which files are present.

Not sure I follow. Modification won't allow files to be deleted, so when `ModB` is installed there will be a superset of the files in `ModA` present on the system. Because `ModB` implicitly `provides` `ModA`, `ModX` should be able to satisfy its `depends` with `ModB`.

**Comment by plague006** [comment]
But mod B =/= mod A.

Ok let's say mod A provides awesomestuff.dll which makes rockets fly upside-down. Mod B modifies awesomestuff.dll so that rockets fly sideways. Mod B can't accurately provide mod A and mods that depend mod A would be borked if they get mod B's version of awesomestuff.dll

**Comment by dbent** [comment]
Ah, now I understand. I should have mentioned that in the OP.

While it's true that a modification could drastically alter the behavior so that mods that `depends` `ModA` would fail to work if `ModB` is installed, I think such cases are likely very rare, and in fact I can't think or an example where that would occur.

*However*, a simple "fix" would be to not have an implicit `provides`. That way if the modifies mod is truly incompatible with the original, we simply don't have it `provides` the original. We'd have to remember to add the explicit `provides` relationship for the other cases though.

(The implicit `conflicts` should always exist though.)

**Comment by dbent** [comment]
And yes, in that case `ModX` would not be installable because it would `depends` `ModA` which `conflcits` with `ModB`. That seems like reasonable behavior to me.

**Comment by plague006** [comment]
You mentioned and I agree: I can't think of a case we have that would cause the issue, and it may never even arise. I just wanted it on the table as an issue that may arise down the road.

**Comment by pjf** [comment]
Still reading the history here, but I'm metioning #949 (saner overwrites) because there's going to be a *lot* of overlap. I'm also going to create milestone for this, as it'll almost certainly require multiple steps.

The puzzles I normally provide are:

---

**Uninstall puzzle**

* User installs A.
* User installs B which overwrites A.
* User uninstalls B.

A is now incomplete. Bonus points: A is no longer in the cache, and the download link is broken.

**Upgrade puzzle**

* User installs A
* User installs B which overwrites A
* User upgrades A

Should we have A's version of files, or B's version?

**Bonus upgrade puzzle**

* User installs A
* User installs B which overwrites A
* User upgrades B which now *no longer* overwrites parts of A

Similar to the first puzzle, how do we get those original files back?

---

My understanding of the proposal is that the relationship would mean that 'B' implicitly conflicts (and may provide) A as part of the scenario, so it's impossible to uninstall B without also uninstalling A. What do we do in the case where an upgrade of A becomes available? What do we do about mods which depend upon a particular version of A, but now we don't have A "installed" as far as the CKAN is concerned?

The CKAN is not-so-secretly an implementation of the Debian spec, which *does* have a [replaces relationship](https://www.debian.org/doc/debian-policy/ch-relationships.html#s-replaces). In this case the "replacing" mod becomes the new owner of those files. This is nice because it clear what we should do (only change those files if the current owner changes them), and it preserves all the other relationships that might exist with the base mod. However it fails terribly if the mod doing the replacing is uninstalled, as the original files may no longer be available. One terrible but possibly effective solution might be to insist the base mod exist in the cache before anything that replaces it can be upgraded/removed. 

The good news is that regardless of what happens we still shouldn't need an installation order. In all cases when installing fileswe can:

* Write the file, if it does not exist
* Decline to overwrite, if it's owned by a mod that overwrites us
* Overwrite the file, if it's owned by a mod we overwrite

We still need to check for circular replacement, but that's some easy work with a graph library.

**Comment by pjf** [comment]
[Milestone created](https://github.com/KSP-CKAN/CKAN/milestones/File%20Overwrites). It's almost certainly incomplete as far as prior discussions go, but I've managed to put all the relevant tickets I could find into it. Feel free to expand as required. :)

**Comment by plague006** [comment]
I actually see your 3 puzzles as one with installation ordering. I think the coding of the relationship has to be such that whenever 1 in the relationship is touched, the following basically happens:

0) Check if downloads of A and B are present.
1) `ModB` gets uninstalled
2) `ModA` gets uninstalled
3) `ModA` gets installed
4) `ModB` gets installed

**Uninstall puzzle** exit the cycle after 3.
**Upgrade puzzle** Full cycle, user should always get `ModB`'s version. If compatibility is broken then `ModB` needs fixing. (That's no different than how things are now, if a dependency breaks or changes, we have to fix it).
**Bonus upgrade puzzle** Full cycle.

You bonus uninstall question is a problem in and of itself. I personally see that as B not being able to have its dependencies met, which is no different than what we have now.

>an upgrade of A becomes available?

Same as dependencies, we rely on mod authors to inform us/give us correct compatibility. We can force (or should be able to.. don't recall if it's working) specific versions to be installed to meet a dependency and this is no different.

----

The tl;dr for me is that every problem presented with `modifies` exists with `depends`. If links break: we're [kinda SOL](KSP-CKAN/NetKAN#932) with `depends`. If an upgrade breaks compatibility we're in the same position as `depends`. Every other issue is just a matter of careful coding (easy for me to say of course because I won't be writing a line of it :smile:)

**Comment by plague006** [comment]
I'll also mention, by making use of the Description column in the changeset tab, we can inform users more clearly of what effects will occur (so that if they lose configs, or compatibility, or whatever they're not totally in the dark). Ex. ![Picture of a stuntman on fire playing the trombone](http://i.imgur.com/dMEJwK6.jpg)

**Comment by dbent** [comment]
I think my proposal is actually much simpler than the impression I gave, hopefully the explanations of the puzzles will clarify how I intend this to work.

At a high level the intent is to basically create "dynamic forks". The use case is to handle the situation where the installation instructions of a mod are:
- Install Mod X
- Install Mod Y, overwriting any existing files

What the modder of Mod Y is effectively doing is *forking* Mod X, without actually going through the trouble of making an actual fork. So we "dynamically" create the fork by making those installation instructions explicit. Outside of the installation/uninstallation process, Mod X and Mod Y should be treated as if they were completely separate mods, with no special relationship. In fact, the `modifies` relationship should be completely invisible to the user, its main goal is to simplify the metadata and help *us*, not the user.

----

For the puzzles:

**Uninstall puzzle**

* User installs A.
  *Mod A is installed as normal*

* User installs B which overwrites A.
  *Bzzt, sorry, `ERROR: Mod B conflicts with Mod A`. If the user wants to use Mod A instead of Mod B, they must uninstall Mod A then install Mod B, as if they were two completely independent mods which had a `conflicts` relationship. Again, `modifies` should be user invisible, it should be as if Mod B was just a fork of Mod A.*

* User uninstalls B.
  *This shouldn't be possible as the previous step would produce an error.*

A is now incomplete. Bonus points: A is no longer in the cache, and the download link is broken. *Shouldn't be possible to get into this situation.*

**Upgrade puzzle**

* User installs A
  *Mod A is installed as normal.*

* User installs B which overwrites A
  *Again, not possible, due to implicit `conflicts`.*

* User upgrades A
  *User can't get here.*

Should we have A's version of files, or B's version?

**Bonus upgrade puzzle**

* User installs A
  *Mod A is installed as normal.*

* User installs B which overwrites A
  *Again, not possible, due to implicit `conflicts`.*

* User upgrades B which now *no longer* overwrites parts of A
  *User can't get here.*

Similar to the first puzzle, how do we get those original files back?

----

So since I answered most of the puzzles with "this can't happen, don't worry about it", how do I actually envisage this working?

**Install `NotSoAwesomeMod`**
  - We download the latest version of `AwesomeMod` depending on version constraints in the `modifies` relationship.
  - We download the latest version of `NotSoAwesomeMod`.
  - We install `AwesomeMod` files.
  - We install `NotSoAwesomeMod` files, overwriting as necessary.
  - The set of files from both `AwesomeMod` and `NotSoAwesomeMod` are considered owned by `NotSoAwesomeMod`, `AwesomeMod` is not considered "installed" at all.

**Upgrade `AwesomeMod`**
Impossible, `AwesomeMod` isn't considered installed, it can't be upgraded.

**Upgrade `NotSoAwesomeMod`**
- We remove all files owned by `NotSoAwesomeMod` (which includes those which would otherwise be owned by `AwesomeMod`)
- We continue as if we were installing `NotSoAwesomeMod` for the first time, again fetching the latest applicable version of `AwesomeMod` files. In this way even though `NotSoAwesomeMod` might not itself have an update available, it will be "upgraded" because its files which really come from `AwesomeMod` will be upgraded. So in a sense the, a mod which has a `modifies` relationship actually has multiple versions: it has its own version and then it has the versions of each mod it `modifies`.

**Uninstall `NotSoAwesomeMod`**
- We remove all files owned by `NotSoAwesomeMod` (which includes those which would otherwise be owned by `AwesomeMod`)

----
> The good news is that regardless of what happens we still shouldn't need an installation order. In all cases when installing files we can: <snip>

True enough.

**Comment by plague006** [comment]
If there's a new release of `AwesomeMod` that is compatible with `NotSoAwesomeMod`, but there's no new release to `NotSoAwesomeMod` how would we get an update to users? Would we have to manually give `NotSoAwesomeMod` a 'virtual' version bump?

Imagine for the sake of argument the `AwesomeMod` upgrade fixes important bugs that aren't addressed by `NotSoAwesomeMod`.

**Comment by dbent** [comment]
That's part of the bit about a mod with `modifies` really having multiple versions. We'd have to keep track of the version of `NotSoAwsomeMod` we have installed *and* the version of `AwesomeMod` we used to install it. Then the upgrade state of `NotSoAwesomeMod` would be determined by the upgrade state of itself and any of its `modifies`.

**Comment by plague006** [comment]
Hardcore. :+1:

**Comment by pjf** [comment]
@dbent: I like your proposal, and thank you for the extra examples, they really help. I'd support a change which added it. You rock.

In the case where people expect `NotSoAwesomeMod` to provide `AwesomeMod`, we can just have an honest-to-goodness `provides` relationship as you've mentioned.

I agree with @plague006 that figuring out when to do upgrades will be tricky, as would displaying to the user modules we used to make up `NotSoAwesomeMod`. We *could* require that `modifies` specify a concrete version of what it works with; that would simply a lot of the handling, but means users can't as easily upgrade when a base mod is upgraded, and means we'd see a lot of funny version numbers to make sure upgrades work the right way.

As an interesting aside, this is isomorphically equivalent the Debian "best practice" for a mod, where it replaces the base mod in its entirety. :)
